index.js

```diff
@@ -637,10 +666,47 @@ SendStream.prototype.sendFile = function sendFile(path) {
     fs.stat(p, function (err, stat) {
       if (err) return next(err)
       if (stat.isDirectory()) return next()
-      self.emit('file', p, stat)
-      self.send(p, stat)
+      checkPrecompressionAndSendFile(p, stat)
     })
   }
+
+  function checkPrecompressionAndSendFile(p, stat) {
+    self.emit('file', p, stat)
+    if (!self._precompressionFormats) return self.send(p, stat)
+
+    var state = {
+      contents: [],
+      extensionsToCheck: self._precompressionFormats.length
+    }
+
+    self._precompressionFormats.forEach(function (format) {
+      debug('stat "%s%s"', p, format.extension);
+      fs.stat(p + format.extension, function onstat(err, contentStat) {
```

Comment:
Hi @gmokki, I'm not saying to not stat at all. I'm saying why stat for all of them. In your example:

> First client sends request without accept-encoding (aka uncompressed).

You don't need to stat at all in this case. Always send the `Vary` header, because the file system can change at any time and a `.gz` file can later appear when it was not there before (think if someone combines this with background compressing on-demand). If you don't expect the file system to change, then you would be caching all these looks, but you're not.